<h1>Activity 1 - Palmer Penguins Data Analytics Project</h1> 
<hr>
<h3> Analyst: Kim David Penaredonda</h3>

In [6]:
#import libraries 
import numpy as np 
import pandas as pd 

ModuleNotFoundError: No module named 'numpy'

<h1>  Converting CSV into a Data Frame </h1>

In [ ]:
penguins = pd.read_csv('data_set//penguins.csv')
penguins

NameError: name 'pd' is not defined

<h1>Describing Data</h1>

In [ ]:
penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      342.000000     342.000000         342.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.459584       1.974793          14.061714   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.225000      15.600000         190.000000  3550.000000   
50%         44.450000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

In [ ]:
penguins.describe().T#transpose 

count         mean         std     min       25%      50%  \
bill_length_mm     342.0    43.921930    5.459584    32.1    39.225    44.45   
bill_depth_mm      342.0    17.151170    1.974793    13.1    15.600    17.30   
flipper_length_mm  342.0   200.915205   14.061714   172.0   190.000   197.00   
body_mass_g        342.0  4201.754386  801.954536  2700.0  3550.000  4050.00   
year               344.0  2008.029070    0.818356  2007.0  2007.000  2008.00   

                      75%     max  
bill_length_mm       48.5    59.6  
bill_depth_mm        18.7    21.5  
flipper_length_mm   213.0   231.0  
body_mass_g        4750.0  6300.0  
year               2009.0  2009.0

<h1> Checkcing Null Values</h1>

In [ ]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

<h1>Checking Value Counts</h1>

In [ ]:
penguins['species']

0         Adelie
1         Adelie
2         Adelie
3         Adelie
4         Adelie
         ...    
339    Chinstrap
340    Chinstrap
341    Chinstrap
342    Chinstrap
343    Chinstrap
Name: species, Length: 344, dtype: object

In [ ]:
penguins['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

<h1>Filtering Data</h1>

In [ ]:
adelie_DF = penguins[penguins['species'] == "Adelie"]
adelie_DF

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
147  Adelie      Dream            36.6           18.4              184.0   
148  Adelie      Dream            36.0           17.8              195.0   
149  Adelie      Dream            37.8           18.1              193.0   
150  Adelie      Dream            36.0           17.1              187.0   
151  Adelie      Dream            41.5           18.5              201.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
147       3475.0  female  2009  
148       3450.0  female  2009  
149       3750.0    male  2009  
150       3700.0  female  2009  
151       4000.0    male  2009  

[152 rows x 8 columns]

In [ ]:
gentoo_DF = penguins[penguins['species'] == "Gentoo"]
gentoo_DF

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
152  Gentoo  Biscoe            46.1           13.2              211.0   
153  Gentoo  Biscoe            50.0           16.3              230.0   
154  Gentoo  Biscoe            48.7           14.1              210.0   
155  Gentoo  Biscoe            50.0           15.2              218.0   
156  Gentoo  Biscoe            47.6           14.5              215.0   
..      ...     ...             ...            ...                ...   
271  Gentoo  Biscoe             NaN            NaN                NaN   
272  Gentoo  Biscoe            46.8           14.3              215.0   
273  Gentoo  Biscoe            50.4           15.7              222.0   
274  Gentoo  Biscoe            45.2           14.8              212.0   
275  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  year  
152       4500.0  female  2007  
153       5700.0    male  2007  
154       4450.0  female  2007  
155       5700.0    male  2007  
156       5400.0    male  2007  
..           ...     ...   ...  
271          NaN     NaN  2009  
272       4850.0  female  2009  
273       5750.0    male  2009  
274       5200.0  female  2009  
275       5400.0    male  2009  

[124 rows x 8 columns]

In [ ]:
chinstrap_DF = penguins[penguins['species'] == "Chinstrap"]
chinstrap_DF

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
276  Chinstrap  Dream            46.5           17.9              192.0   
277  Chinstrap  Dream            50.0           19.5              196.0   
278  Chinstrap  Dream            51.3           19.2              193.0   
279  Chinstrap  Dream            45.4           18.7              188.0   
280  Chinstrap  Dream            52.7           19.8              197.0   
..         ...    ...             ...            ...                ...   
339  Chinstrap  Dream            55.8           19.8              207.0   
340  Chinstrap  Dream            43.5           18.1              202.0   
341  Chinstrap  Dream            49.6           18.2              193.0   
342  Chinstrap  Dream            50.8           19.0              210.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
276       3500.0  female  2007  
277       3900.0    male  2007  
278       3650.0    male  2007  
279       3525.0  female  2007  
280       3725.0    male  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[68 rows x 8 columns]

<h1>Questions</h1>
<hr>
<p>How many of these Male Chinstrap Species weight is above 400grams?</p>

In [ ]:
heavy_chinstrap = penguins[(penguins['species'] == 'Chinstrap') & (penguins['body_mass_g'] >= 4000)]

In [ ]:
heavy_chinstrap

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
284  Chinstrap  Dream            46.0           18.9              195.0   
289  Chinstrap  Dream            52.0           18.1              201.0   
291  Chinstrap  Dream            50.5           19.6              201.0   
295  Chinstrap  Dream            49.2           18.2              195.0   
301  Chinstrap  Dream            52.0           19.0              197.0   
305  Chinstrap  Dream            52.8           20.0              205.0   
307  Chinstrap  Dream            54.2           20.8              201.0   
309  Chinstrap  Dream            51.0           18.8              203.0   
313  Chinstrap  Dream            52.0           20.7              210.0   
315  Chinstrap  Dream            53.5           19.9              205.0   
321  Chinstrap  Dream            50.8           18.5              201.0   
323  Chinstrap  Dream            49.0           19.6              212.0   
329  Chinstrap  Dream            50.7           19.7              203.0   
333  Chinstrap  Dream            49.3           19.9              203.0   
339  Chinstrap  Dream            55.8           19.8              207.0   
342  Chinstrap  Dream            50.8           19.0              210.0   

     body_mass_g     sex  year  
284       4150.0  female  2007  
289       4050.0    male  2007  
291       4050.0    male  2007  
295       4400.0    male  2007  
301       4150.0    male  2007  
305       4550.0    male  2008  
307       4300.0    male  2008  
309       4100.0    male  2008  
313       4800.0    male  2008  
315       4500.0    male  2008  
321       4450.0    male  2009  
323       4300.0    male  2009  
329       4050.0    male  2009  
333       4050.0    male  2009  
339       4000.0    male  2009  
342       4100.0    male  2009

<h1>Filling Up Null Values</h1>

In [ ]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [ ]:
average_bill_len = penguins['bill_length_mm'].mean()
penguins['bill_length_mm'] = penguins['bill_length_mm'].fillna(average_bill_len)
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [ ]:
average_bill_depth = penguins['bill_depth_mm'].mean()
penguins['bill_depth_mm'] = penguins['bill_depth_mm'].fillna(average_bill_depth)
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [ ]:
average_flipper_ave = penguins['flipper_length_mm'].mean()
penguins['flipper_length_mm'] = penguins['flipper_length_mm'].fillna(average_flipper_ave)
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     0
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [ ]:
average_body_mass = penguins['body_mass_g'].mean()

In [ ]:
penguins['body_mass_g'] = penguins['body_mass_g'].fillna(average_body_mass)
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     0
body_mass_g           0
sex                  11
year                  0
dtype: int64

In [ ]:
frequency_sex = penguins['sex'].mode()
penguins['sex'] = penguins['sex'].fillna(frequency_sex[0])
penguins.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

In [ ]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     344 non-null    float64
 3   bill_depth_mm      344 non-null    float64
 4   flipper_length_mm  344 non-null    float64
 5   body_mass_g        344 non-null    float64
 6   sex                344 non-null    object 
 7   year               344 non-null    int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 21.6+ KB


In [ ]:
penguins['species'] = penguins['species'].astype('category')
penguins['island'] = penguins['island'].astype('category') 
penguins['sex'] = penguins['sex'].astype('category')
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    category
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      344 non-null    float64 
 4   flipper_length_mm  344 non-null    float64 
 5   body_mass_g        344 non-null    float64 
 6   sex                344 non-null    category
 7   year               344 non-null    int64   
dtypes: category(3), float64(4), int64(1)
memory usage: 15.0 KB


<h1> Exploratory Data Analysis</h1>

<h1>1. What is the highest body mass of all penguin Species?</h1>

In [ ]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen        39.10000       18.70000         181.000000   
1       Adelie  Torgersen        39.50000       17.40000         186.000000   
2       Adelie  Torgersen        40.30000       18.00000         195.000000   
3       Adelie  Torgersen        43.92193       17.15117         200.915205   
4       Adelie  Torgersen        36.70000       19.30000         193.000000   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream        55.80000       19.80000         207.000000   
340  Chinstrap      Dream        43.50000       18.10000         202.000000   
341  Chinstrap      Dream        49.60000       18.20000         193.000000   
342  Chinstrap      Dream        50.80000       19.00000         210.000000   
343  Chinstrap      Dream        50.20000       18.70000         198.000000   

     body_mass_g     sex  year  
0    3750.000000    male  2007  
1    3800.000000  female  2007  
2    3250.000000  female  2007  
3    4201.754386    male  2007  
4    3450.000000  female  2007  
..           ...     ...   ...  
339  4000.000000    male  2009  
340  3400.000000  female  2009  
341  3775.000000    male  2009  
342  4100.000000    male  2009  
343  3775.000000  female  2009  

[344 rows x 8 columns]

In [ ]:
penguins["species"].unique()

['Adelie', 'Gentoo', 'Chinstrap']
Categories (3, object): ['Adelie', 'Chinstrap', 'Gentoo']

In [ ]:
np.array(['Adelie', 'Gentoo', 'Chinstrap'], dtype=object)

array(['Adelie', 'Gentoo', 'Chinstrap'], dtype=object)

In [ ]:
avg_body_mass = penguins.groupby("species")["body_mass_g"].mean().round(2)

In [ ]:
avg_body_mass

species
Adelie       3703.96
Chinstrap    3733.09
Gentoo       5068.97
Name: body_mass_g, dtype: float64

<h1>INSIGHT</h1>
<hr>
<p> Based on its outcome we can see that the Gentoo penguins have the highest body mass compared to other specie of penguins.</p>

<h1>2. What specie have the longest flipper length?</h1>

In [ ]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen        39.10000       18.70000         181.000000   
1       Adelie  Torgersen        39.50000       17.40000         186.000000   
2       Adelie  Torgersen        40.30000       18.00000         195.000000   
3       Adelie  Torgersen        43.92193       17.15117         200.915205   
4       Adelie  Torgersen        36.70000       19.30000         193.000000   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream        55.80000       19.80000         207.000000   
340  Chinstrap      Dream        43.50000       18.10000         202.000000   
341  Chinstrap      Dream        49.60000       18.20000         193.000000   
342  Chinstrap      Dream        50.80000       19.00000         210.000000   
343  Chinstrap      Dream        50.20000       18.70000         198.000000   

     body_mass_g     sex  year  
0    3750.000000    male  2007  
1    3800.000000  female  2007  
2    3250.000000  female  2007  
3    4201.754386    male  2007  
4    3450.000000  female  2007  
..           ...     ...   ...  
339  4000.000000    male  2009  
340  3400.000000  female  2009  
341  3775.000000    male  2009  
342  4100.000000    male  2009  
343  3775.000000  female  2009  

[344 rows x 8 columns]

In [ ]:
penguins[["flipper_length_mm", "species"]].value_counts()

flipper_length_mm  species  
190.0              Adelie       19
215.0              Gentoo       12
195.0              Adelie       11
187.0              Adelie       10
193.0              Adelie       10
                                ..
205.0              Adelie        1
207.0              Chinstrap     1
206.0              Chinstrap     1
226.0              Gentoo        1
231.0              Gentoo        1
Name: count, Length: 84, dtype: int64

In [ ]:
flipper_averages = (
    'Adelie:', round(float(penguins[penguins["species"] == "Adelie"]["flipper_length_mm"].mean()), 2),
    'Chinstrap:', round(float(penguins[penguins["species"] == "Chinstrap"]["flipper_length_mm"].mean()), 2),
    'Gentoo:', round(float(penguins[penguins["species"] == "Gentoo"]["flipper_length_mm"].mean()), 2)
)

In [ ]:
flipper_averages

('Adelie:', 190.03, 'Chinstrap:', 195.82, 'Gentoo:', 217.06)

<h3>INSIGHT</h3>
<hr>
<p>The species of penguin that has the longest flipper length is Gentoo compared to other species of penguin</p>

<h1>3. In what year has the heaviest penguin?</h1>

In [ ]:
max_bmg = penguins['body_mass_g'].max()
print(max_bmg)
penguins[(penguins['year']) & (penguins['body_mass_g'] >= max_bmg)]

6300.0


species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
169  Gentoo  Biscoe            49.2           15.2              221.0   

     body_mass_g   sex  year  
169       6300.0  male  2007

<h1>INSIGHT</h1>
<hr>
<p>The graph shows the heaviest penguin ever recorded was in 2007.</p>

<h1>4.What is the average bill length for each specie?</h1>

In [ ]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen        39.10000       18.70000         181.000000   
1       Adelie  Torgersen        39.50000       17.40000         186.000000   
2       Adelie  Torgersen        40.30000       18.00000         195.000000   
3       Adelie  Torgersen        43.92193       17.15117         200.915205   
4       Adelie  Torgersen        36.70000       19.30000         193.000000   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream        55.80000       19.80000         207.000000   
340  Chinstrap      Dream        43.50000       18.10000         202.000000   
341  Chinstrap      Dream        49.60000       18.20000         193.000000   
342  Chinstrap      Dream        50.80000       19.00000         210.000000   
343  Chinstrap      Dream        50.20000       18.70000         198.000000   

     body_mass_g     sex  year  
0    3750.000000    male  2007  
1    3800.000000  female  2007  
2    3250.000000  female  2007  
3    4201.754386    male  2007  
4    3450.000000  female  2007  
..           ...     ...   ...  
339  4000.000000    male  2009  
340  3400.000000  female  2009  
341  3775.000000    male  2009  
342  4100.000000    male  2009  
343  3775.000000  female  2009  

[344 rows x 8 columns]

In [ ]:
penguins[["bill_length_mm", "species"]].value_counts()

bill_length_mm  species  
41.1            Adelie       7
37.8            Adelie       5
39.6            Adelie       5
39.7            Adelie       4
38.1            Adelie       4
                            ..
55.1            Gentoo       1
55.8            Chinstrap    1
55.9            Gentoo       1
58.0            Chinstrap    1
59.6            Gentoo       1
Name: count, Length: 210, dtype: int64

<h1>INSIGHT</h1>
<hr
<p>This shows that the Chinstrap Specie has the longest bill length among species.</p>

In [ ]:
# adelie_data

<h1>5. What sex have the shortest bill length?</h1>

In [ ]:
 penguins['bill_length_mm'].min()

np.float64(32.1)

In [ ]:
penguins[penguins['bill_length_mm'] == 32.1]

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
142  Adelie  Dream            32.1           15.5              188.0   

     body_mass_g     sex  year  
142       3050.0  female  2009

<h1>Insight</h1>
<hr>
<p>Based on this, Adelie females have the shortest bill length, with the minimum recorded being 32.1 mm</p>

<h1>6.Which island has the most specie of Gentoo penguins?</h1>

In [ ]:
penguins[penguins['species'] == 'Gentoo']['island'].value_counts()

island
Biscoe       124
Dream          0
Torgersen      0
Name: count, dtype: int64

<h1>INSIGHT</h1>
<hr>
<p>This graph shows that gentoo penguins mostly live in biscoe islands.
</p>

<h1>7. What is the lightest body mass of female adelie in the year 2008?</h1>

In [ ]:
female_adelie_2008 = penguins[(penguins["species"] == "Adelie") & (penguins["sex"] == "female") & (penguins["year"] == 2008)]
lightest_mass = female_adelie_2008["body_mass_g"].min()

In [ ]:
female_adelie_2008

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
50  Adelie     Biscoe            39.6           17.7              186.0   
52  Adelie     Biscoe            35.0           17.9              190.0   
54  Adelie     Biscoe            34.5           18.1              187.0   
56  Adelie     Biscoe            39.0           17.5              186.0   
58  Adelie     Biscoe            36.5           16.6              181.0   
60  Adelie     Biscoe            35.7           16.9              185.0   
62  Adelie     Biscoe            37.6           17.0              185.0   
64  Adelie     Biscoe            36.4           17.1              184.0   
66  Adelie     Biscoe            35.5           16.2              195.0   
68  Adelie  Torgersen            35.9           16.6              190.0   
70  Adelie  Torgersen            33.5           19.0              190.0   
72  Adelie  Torgersen            39.6           17.2              196.0   
74  Adelie  Torgersen            35.5           17.5              190.0   
76  Adelie  Torgersen            40.9           16.8              191.0   
78  Adelie  Torgersen            36.2           16.1              187.0   
80  Adelie  Torgersen            34.6           17.2              189.0   
82  Adelie  Torgersen            36.7           18.8              187.0   
84  Adelie      Dream            37.3           17.8              191.0   
87  Adelie      Dream            36.9           18.6              189.0   
89  Adelie      Dream            38.9           18.8              190.0   
90  Adelie      Dream            35.7           18.0              202.0   
92  Adelie      Dream            34.0           17.1              185.0   
94  Adelie      Dream            36.2           17.3              187.0   
96  Adelie      Dream            38.1           18.6              190.0   
98  Adelie      Dream            33.1           16.1              178.0   

    body_mass_g     sex  year  
50       3500.0  female  2008  
52       3450.0  female  2008  
54       2900.0  female  2008  
56       3550.0  female  2008  
58       2850.0  female  2008  
60       3150.0  female  2008  
62       3600.0  female  2008  
64       2850.0  female  2008  
66       3350.0  female  2008  
68       3050.0  female  2008  
70       3600.0  female  2008  
72       3550.0  female  2008  
74       3700.0  female  2008  
76       3700.0  female  2008  
78       3550.0  female  2008  
80       3200.0  female  2008  
82       3800.0  female  2008  
84       3350.0  female  2008  
87       3500.0  female  2008  
89       3600.0  female  2008  
90       3550.0  female  2008  
92       3400.0  female  2008  
94       3300.0  female  2008  
96       3700.0  female  2008  
98       2900.0  female  2008

In [ ]:
lightest_mass

NameError: name 'lightest_mass' is not defined

<h1>INSIGHT</h1>
<hr>
<p>The result shows that the lightest mass of adelie specie in year 2008 is 2.8kg body mass.</p>

<h1>Data Aggregation</h1>


<h1>8) Which sex of Gentoo penguins has the highest average body mass?</h1>


In [ ]:
penguins.groupby('sex')['flipper_length_mm'].max()

NameError: name 'penguins' is not defined

<h1>insight</h1>

<h1>9) Which species has the maximum flipper length</h1>

In [ ]:
penguins.groupby('species')['flipper_length_mm'].max()


penguins['flipper_length_mm'].max()

<h1>insight</h1>

<h1>10)  What is the minimum and maximum flipper length per group</h1>

In [ ]:
penguins.groupby(['species', 'island'])['flipper_length_mm'].agg(['min', 'max'])

<h1>insight</h1>

<h1>11)How many penguins of each species live on each island??</h1>

In [ ]:

species_island = penguins.groupby(['species', 'island']).count()
species_island


<h1>Insight No.9</h1>
<p>Adelie penguins are present on Biscoe (44), Dream (56), and Torgersen (52) Islands; Chinstrap penguins are only on Dream Island (68); and Gentoo penguins are found exclusively on Biscoe Island (124).</p>

<h1>12) Give the yearly average mass for each species, island, and sex?</h1>

In [ ]:
penguins.groupby(['species', 'island', 'sex', 'year'])['mass_kg'].mean()

<h1>insight</h1>

<h1>13)Is there a difference in the bill length between the three penguin species?</h1>

In [ ]:
bill_length, p_value = f_oneway(penguins[penguins['species'] == 'Adelie']['bill_length_mm'], 
                           penguins[penguins['species'] == 'Chinstrap']['bill_length_mm'], 
                           penguins[penguins['species'] == 'Gentoo']['bill_length_mm'])

In [ ]:

p_value

<h1>insight</h1>